zipkin-server/src/test/java/zipkin2/server/internal/eureka/ITZipkinEureka.java

```diff
@@ -0,0 +1,128 @@
+import zipkin2.server.internal.ZipkinConfiguration;
+import zipkin2.server.internal.ZipkinHttpConfiguration;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.awaitility.Awaitility.await;
+
+public class ITZipkinEureka {
+
+  static final AtomicReference<HttpData> registerContentCaptor = new AtomicReference<>();
+
+  static final CompletableFuture<RequestHeaders> heartBeatHeadersCaptor =
+    new CompletableFuture<>();
+  static final CompletableFuture<RequestHeaders> deregisterHeadersCaptor =
+    new CompletableFuture<>();
+
+  @RegisterExtension static final ServerExtension eurekaServer = new ServerExtension() {
+    @Override protected void configure(ServerBuilder sb) {
+      sb.service("/apps/zipkin", (ctx, req) -> {
+        if (req.method() != HttpMethod.POST) {
+          registerContentCaptor.set(null);
+          return HttpResponse.of(HttpStatus.METHOD_NOT_ALLOWED);
+        }
+        final CompletableFuture<HttpResponse> future = new CompletableFuture<>();
+        req.aggregate().handle((aggregatedRes, cause) -> {
+          registerContentCaptor.set(aggregatedRes.content());
+          future.complete(HttpResponse.of(HttpStatus.NO_CONTENT));
+          return null;
+        });
+        return HttpResponse.from(future);
+      });
+      final AtomicInteger heartBeatRequestCounter = new AtomicInteger();
+      sb.service("/apps/zipkin/{instanceId}", (ctx, req) -> {
+        req.aggregate();
+        if (req.method() == HttpMethod.PUT) {
+          final int count = heartBeatRequestCounter.getAndIncrement();
+          if (count == 0) {
+            // This is for the test that EurekaUpdatingListener automatically retries when
+            // RetryingClient is not used.
+            return HttpResponse.of(HttpStatus.INTERNAL_SERVER_ERROR);
+          }
+          heartBeatHeadersCaptor.complete(req.headers());
+        } else if (req.method() == HttpMethod.DELETE) {
+          deregisterHeadersCaptor.complete(req.headers());
+        }
+        return HttpResponse.of(HttpStatus.OK);
+      });
+    }
+  };
+
+  AnnotationConfigApplicationContext context = new AnnotationConfigApplicationContext();
+  Server server;
+
+  @BeforeEach void init() {
+    TestPropertyValues.of(
+      "server.port=9413",
+      "zipkin.query.enabled=false",
+      "zipkin.ui.enabled=false",
+      "zipkin.discovery.eureka.url=" + eurekaServer.httpUri()
+    ).applyTo(context);
+
+    context.register(
+      ConversionServiceConfig.class,
+      PropertyPlaceholderAutoConfiguration.class,
+      ArmeriaAutoConfiguration.class,
+      ZipkinConfiguration.class,
+      ZipkinHttpConfiguration.class,
+      ZipkinEurekaConfiguration.class
+    );
+    context.refresh();
+
+    server = context.getBean(Server.class);
+  }
+
+  @Test void registersInEureka() {
+    await().until(() -> registerContentCaptor.get() != null);
+    assertThat(registerContentCaptor.get().toStringAscii())
+      .isEqualTo("sdd");
```

Comment:
this shows the json isn't sane

Reply:
https://github.com/openzipkin/zipkin/pull/3316/checks?check_run_id=1440795186#step:5:4120
So this is the content of the registering request. Which one is wrong?

Reply:
ah.. so I had to hard-code the port 9413 in order for it to show up here.. also I suspect the secure port 0 should not show up (not sure their protocol, but seems better to omit than say port 0)

the main issue is around default of the port and being able to indicate a healthcheck path as opposed to having to construct it.

```json
{"instance":{"instanceId":"zipkin-server","app":"zipkin","hostName":"fv-az75-872","ipAddr":"10.1.0.4","vipAddress":"fv-az75-872:9413","port":{"$":9413,"enabled":"true"},"securePort":{"$":0,"enabled":"false"},"status":"UP","healthCheckUrl":"http://localhost:9413/health","dataCenterInfo":{"@class":"com.netflix.appinfo.InstanceInfo$DefaultDataCenterInfo","name":"MyOwn","metadata":{}},"leaseInfo":{"renewalIntervalInSecs":30,"durationInSecs":90},"metadata":{},"lastUpdatedTimestamp":1606118031032,"lastDirtyTimestamp":1606118031032}}
```

Reply:
IIRC, it needs to guarantee compatibility with the legacy. 🤔 
https://github.com/Netflix/eureka/blob/07876a70a547ea52822c286b49c001971a1e6880/eureka-client/src/main/java/com/netflix/discovery/converters/jackson/serializer/InstanceInfoJsonBeanSerializer.java#L56